src/masternodes/rpc_icxorderbook.cpp

```diff
@@ -436,68 +479,124 @@ UniValue icxsubmitdfchtlc(const JSONRPCRequest& request) {
+    else
+        throw JSONRPCError(RPC_INVALID_PARAMETER,"Invalid parameters, argument \"receiveAddress\" must be non-null");
+    if (!metaObj["hash"].isNull())
+        submitdfchtlc.hash = uint256S(metaObj["hash"].getValStr());
+    else
+        throw JSONRPCError(RPC_INVALID_PARAMETER,"Invalid parameters, argument \"hash\" must be non-null");
+    if (!metaObj["timeout"].isNull())
         submitdfchtlc.timeout = metaObj["timeout"].get_int();
-    }
 
     int targetHeight;
     {
         LOCK(cs_main);
         auto offer = pcustomcsview->GetICXMakeOfferByCreationTx(submitdfchtlc.offerTx);
         if (!offer)
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "offerTx (" + submitdfchtlc.offerTx.GetHex() + ") does not exist");
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("offerTx (%s) does not exist",submitdfchtlc.offerTx.GetHex()));
         auto order = pcustomcsview->GetICXOrderByCreationTx(offer->orderTx);
         if (!order)
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "orderTx (" + offer->orderTx.GetHex() + ") does not exist");
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("orderTx (%s) does not exist",offer->orderTx.GetHex()));
 
         if (order->orderType == CICXOrder::TYPE_INTERNAL)
         {
+            if (submitdfchtlc.amount * order->orderPrice / COIN != offer->amount)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("cannot make dfc htlc with that amount, different amount necessary for offer (%s) - %s != %s!",
+                                offer->creationTx.GetHex(), ValueFromAmount(submitdfchtlc.amount * order->orderPrice / COIN).getValStr(), ValueFromAmount(offer->amount).getValStr()));
+
             if (!metaObj["receivePubkey"].isNull()) {
-                submitdfchtlc.receiveDestination = ParseHex(trim_ws(metaObj["receivePubkey"].getValStr()));
+                submitdfchtlc.receivePubkey = PublickeyFromString(trim_ws(metaObj["receivePubkey"].getValStr()));
+                if (!submitdfchtlc.receivePubkey.IsFullyValid())
+                    throw JSONRPCError(RPC_INVALID_PARAMETER,"Invalid parameters, argument \"receivePubKey\" is not a valid pubkey");
             }
-            else throw JSONRPCError(RPC_INVALID_PARAMETER,"Invalid parameters, argument \"receivePubKey\" must be non-null or invalid pubkey");
+            else
+                throw JSONRPCError(RPC_INVALID_PARAMETER,"Invalid parameters, argument \"receivePubKey\" must be non-null");
+
+            bool found = false;
+            pcustomcsview->ForEachICXSubmitDFCHTLCOpen([&found, &submitdfchtlc](CICXOrderView::TxidPairKey const & key, uint8_t i) {
+                if (key.first != submitdfchtlc.offerTx) 
+                    return false;
+                found = true;
+                return false;
+            }, submitdfchtlc.offerTx);
+            if (found)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "dfc htlc already submitted!");
```

Comment:
Make a simple function in the view `bool HasICXSubmitDFCHTLCOpen` in other places as well.